### ISSUE TITLE
BatchPrefillWithRaggedKVCacheWrapper cutlass backend doesn't match FA2 when head_dim=192

### ISSUE THREAD

**Issue by wenscarl** [flashinfer-ai/flashinfer#1663]
On blackwell, `BatchPrefillWithRaggedKVCacheWrapper` cutlass backend doesn't match FA2 backend when head_dim=192, especially when `causal=False`. This causes accuracy issue: https://github.com/sgl-project/sglang/issues/9806
reproducer:
```
# reproducer.py
import torch
import flashinfer

def test_batch_prefill():
    device = "cuda:0"

    # parameters
    num_layers = 2  # keep small for test
    num_qo_heads = 128
    num_kv_heads = 128
    head_dim = 192 # fail
    #head_dim = 127 # pass
    v_head_dim = 128

    # nnz tokens
    nnz_qo = 256
    nnz_kv = 15535
    batch_size = 2  # example with 2 sequences

    # allocate workspace
    workspace_buffer = torch.empty(128 * 1024 * 1024, dtype=torch.uint8, device=device)
    prefill_wrapper = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(workspace_buffer, "NHD", backend="cutlass")
    prefill_wrapper_baseline = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(workspace_buffer, "NHD", backend="auto")

    # create indptr arrays (ragged structure)
    # Example: two sequences, lengths split
    #qo_indptr = torch.tensor([0, nnz_qo], dtype=torch.int32, device=device)
    qo_indptr = torch.tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=torch.int32, device=device)
    #kv_indptr = torch.tensor([0, nnz_kv], dtype=torch.int32, device=device)
    kv_indptr = torch.tensor([0,  1279,  2609,  3885,  5185,  6497,  7783,  9072, 10378, 11661, 12935, 14251, 15531], dtype=torch.int32, device=device)

    # random q/k/v inputs
    q = torch.randn(nnz_qo, num_qo_heads, head_dim, dtype=torch.bfloat16, device=device)
    k = torch.randn(nnz_kv, num_kv_heads, head_dim, dtype=torch.bfloat16, device=device)
    v = torch.randn(nnz_kv, num_kv_heads, v_head_dim, dtype=torch.bfloat16, device=device)

    # plan with causal mask
    prefill_wrapper.plan(
        qo_indptr,
        kv_indptr,
        num_qo_heads,
        num_kv_heads,
        head_dim,
        causal=False,
        q_data_type=torch.bfloat16,
    )
    prefill_wrapper_baseline.plan(
        qo_indptr,
        kv_indptr,
        num_qo_heads,
        num_kv_heads,
        head_dim,
        causal=False,
        q_data_type=torch.bfloat16,
    )

    outputs = []
    o = prefill_wrapper.run(q, k, v)
    o_baseline = prefill_wrapper_baseline.run(q, k, v)
#    print(f"o:{o}; o_baseline:{o_baseline}")
#    print(f"Layer {0} output shape: {o.shape}")
    torch.testing.assert_close(o, o_baseline, rtol=1e-2, atol=1e-2)


if __name__ == "__main__":
    test_batch_prefill()
```
`python reproducer.py` will see mismatch.
@yzh119 @elfiegg

**Comment by yzh119** [comment]
bug confirmed, working on it now.

**Comment by yzh119** [comment]
Hi @wenscarl , `head_dim_vo` is missing from the plan functions, after passing this value to plan, the tests passed.
```python
    # plan with causal mask
    prefill_wrapper.plan(
        qo_indptr,
        kv_indptr,
        num_qo_heads,
        num_kv_heads,
        head_dim,
        causal=False,
        head_dim_vo=v_head_dim,
        q_data_type=torch.bfloat16,
    )
    prefill_wrapper_baseline.plan(
        qo_indptr,
        kv_indptr,
        num_qo_heads,
        num_kv_heads,
        head_dim,
        causal=False,
        head_dim_vo=v_head_dim,
        q_data_type=torch.bfloat16,
    )
```

**Comment by wenscarl** [comment]
The above reproducer is a false alarm. Thanks @elfiegg for providing the following reproducer which can capture the mismatch.
```
@pytest.mark.parametrize("batch_size", [129])
@pytest.mark.parametrize("s_qo", [1])
@pytest.mark.parametrize("s_kv", [1])
@pytest.mark.parametrize("num_kv_heads", [128])
@pytest.mark.parametrize("head_grp_size", [1])
@pytest.mark.parametrize("causal", [True, False])
def test_trtllm_gen_prefill_deepseek(
    batch_size, s_qo, s_kv, num_kv_heads, head_grp_size, causal
):
    if s_qo > s_kv:
        pytest.skip("s_qo > s_kv, skipping test as causal")
    num_qo_heads = num_kv_heads * head_grp_size
    head_dim_qk = 192
    head_dim_vo = 128
    seed = 0
    torch.manual_seed(seed)
    device = "cuda:0"
    # actual_seq_lens_q = torch.randint(
    #     1, s_qo + 1, (batch_size, 1, 1, 1), dtype=torch.int32, device=device
    # )
    actual_seq_lens_q = torch.tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1], dtype=torch.int32, device=device)
    # actual_seq_lens_kv = torch.randint(
    #     s_qo, s_kv + 1, (batch_size, 1, 1, 1), dtype=torch.int32, device=device
    # )
    actual_seq_lens_kv = actual_seq_lens_q
    cumsum_s_qo = torch.sum(actual_seq_lens_q)
    cumsum_s_kv = torch.sum(actual_seq_lens_kv)
    q = torch.randn(
        982, num_qo_heads, head_dim_qk, device=device, dtype=torch.bfloat16
    )
    k_cache = torch.randn(
        (982, num_kv_heads, head_dim_qk),
        device=device,
        dtype=torch.bfloat16,
    )
    v_cache = torch.randn(
        (982, num_kv_heads, head_dim_vo),
        device=device,
        dtype=torch.bfloat16,
    )
    # Initialize scale
    scale = float(1.0 / (head_dim_qk**0.5))
    workspace_buffer = torch.empty(workspace_size, dtype=torch.int8, device=device)
    qo_indptr = torch.cat(
        [
            torch.tensor([0], device=device),
            torch.cumsum(actual_seq_lens_q.view(-1), dim=0),
        ]
    ).int()
    # kv_indptr = torch.arange(0, batch_size + 1, device="cuda", dtype=torch.int32) * s_kv
    # Create kv_indptr as cumulative sum of actual_seq_lens_kv
    kv_indptr = torch.cat(
        [
            torch.tensor(
                [0],
                device=device,
            ),
            torch.cumsum(actual_seq_lens_kv.view(-1), dim=0),
        ]
    ).int()
    wrapper = flashinfer.prefill.BatchPrefillWithRaggedKVCacheWrapper(
        torch.zeros(workspace_size, device="cuda", dtype=torch.uint8),
        kv_layout="NHD",
        backend="fa2",
    )
    wrapper.plan(
        qo_indptr,
        kv_indptr,
        num_qo_heads,
        num_kv_heads,
        head_dim_qk,
        head_dim_vo=head_dim_vo,
        causal=causal,
        sm_scale=scale,
        q_data_type=torch.bfloat16,
        kv_data_type=torch.bfloat16,
    )
    output_ref, lse_ref = wrapper.run(q, k_cache, v_cache, return_lse=True)
    output = torch.empty_like(output_ref)
    bmm1_scale = scale
    bmm2_scale = 1.0
    output_trtllm, lse_trtllm = flashinfer.prefill.trtllm_ragged_attention_deepseek(
        q,
        k_cache,
        v_cache,
        workspace_buffer,
        actual_seq_lens_kv,
        s_qo,
        s_kv,
        bmm1_scale,
        bmm2_scale,
        -1,
        batch_size,
        -1,
        qo_indptr,
        kv_indptr,
        False,
        causal,
        True,
        out=output,
    )
    torch.testing.assert_close(
        output_trtllm,
        output_ref,
        atol=1e-2,
        rtol=1e-2,
    )
    torch.testing.assert_close(
        lse_trtllm,
        lse_ref,
        atol=1e-3,
        rtol=1e-3,
    )
```

**Comment by yzh119** [comment]
For this new example, the main issue is that the first dimension of `q` exceeds `cumsum_s_qo[-1]` (we only guarantee that `output[:cumsum_s_qo[-1]]` are correct) and the behavior of both kernels on out-of-boundary outputs is undefined, what behavior do you expect on oob elements (make them all zero)?

**Comment by wenscarl** [comment]
In the first example where comparing fa2 vs cutlass, the first dim of q also exceeds cumsum_s_qo[-1], both backends set oob positions to 0.

**Comment by wenscarl** [comment]
```
@pytest.mark.parametrize("batch_size", [2])
@pytest.mark.parametrize("s_qo", [1])
@pytest.mark.parametrize("s_kv", [1])
@pytest.mark.parametrize("num_kv_heads", [128])
@pytest.mark.parametrize("head_grp_size", [1])
@pytest.mark.parametrize("causal", [True, False])
def test_trtllm_gen_prefill_deepseek(
    batch_size, s_qo, s_kv, num_kv_heads, head_grp_size, causal
):
    if s_qo > s_kv:
        pytest.skip("s_qo > s_kv, skipping test as causal")
    num_qo_heads = num_kv_heads * head_grp_size
    head_dim_qk = 192
    head_dim_vo = 128
    seed = 0
    torch.manual_seed(seed)
    device = "cuda:0"
    actual_seq_lens_q = torch.tensor([23, 192], dtype=torch.int32, device=device)
    actual_seq_lens_kv = torch.tensor([1280, 0], dtype=torch.int32, device=device)
    q = torch.randn(
        215, num_qo_heads, head_dim_qk, device=device, dtype=torch.bfloat16
    )
    k_cache = torch.randn(
        (1280, num_kv_heads, head_dim_qk),
        device=device,
        dtype=torch.bfloat16,
    )
    v_cache = torch.randn(
        (1280, num_kv_heads, head_dim_vo),
        device=device,
        dtype=torch.bfloat16,
    )
    # Initialize scale
    scale = float(1.0 / (head_dim_qk**0.5))
    workspace_buffer = torch.empty(workspace_size, dtype=torch.int8, device=device)
    qo_indptr = torch.cat(
        [
            torch.tensor([0], device=device),
            torch.cumsum(actual_seq_lens_q.view(-1), dim=0),
        ]
    ).int()
    # kv_indptr = torch.arange(0, batch_size + 1, device="cuda", dtype=torch.int32) * s_kv
    # Create kv_indptr as cumulative sum of actual_seq_lens_kv
    kv_indptr = torch.cat(
        [
            torch.tensor(
                [0],
                device=device,
            ),
            torch.cumsum(actual_seq_lens_kv.view(-1), dim=0),
        ]
    ).int()
    wrapper = flashinfer.prefill.BatchPrefillWithRaggedKVCacheWrapper(
            #torch.zeros(workspace_size, device="cuda", dtype=torch.uint8),
        workspace_buffer,
        kv_layout="NHD",
        backend="cutlass",
    )
    wrapper.plan(
        qo_indptr,
        kv_indptr,
        num_qo_heads,
        num_kv_heads,
        head_dim_qk,
        head_dim_vo=head_dim_vo,
        causal=causal,
        sm_scale=scale,
        q_data_type=torch.bfloat16,
        kv_data_type=torch.bfloat16,
    )
    output = torch.ones(215, num_qo_heads, head_dim_vo, device=q.device, dtype=q.dtype)
    output_ref, lse_ref = wrapper.run(q, k_cache, v_cache, return_lse=True, out=output)
    bmm1_scale = scale
    bmm2_scale = 1.0
    output2 = torch.ones(215, num_qo_heads, head_dim_vo, device=q.device, dtype=q.dtype)
    output_trtllm, lse_trtllm = flashinfer.prefill.trtllm_ragged_attention_deepseek(
        q,
        k_cache,
        v_cache,
        workspace_buffer,
        actual_seq_lens_kv,
        #s_qo,
        #s_kv,
        actual_seq_lens_q.max().item(),
        actual_seq_lens_kv.max().item(),
        bmm1_scale,
        bmm2_scale,
        -1,
        batch_size,
        -1,
        qo_indptr,
        kv_indptr,
        False,
        is_causal=causal,
        return_lse=True,
        #   out=output2,
    )
#    print(lse_trtllm)
#    print(lse_ref)
    print(output_trtllm)
    print(output_ref)
    print(f"bs:{batch_size}")
    print(f"q:{q.shape}")
    print(f"qoind:{qo_indptr[-1]}")
    torch.testing.assert_close(
        output_trtllm,
        output_ref,
        atol=1e-2,
        rtol=1e-2,
    )
```
**For this test, the causal=True passed but causal=False failed.** -< not true.

**Comment by wenscarl** [comment]
By running the above script, the root cause can be summarized as:
1. The output buffer is torch.empty if not provided. 
2. cutlass backend will always zero out oom positions while trtllm backend doesn't. For the usecase in sglang, the output should be reset to zero due to following all-reduce.

closing.
